### ISSUE TITLE
[Question] Can't load ICU library

### ISSUE THREAD

**Issue by funex** [sillsdev/icu-dotnet#30]
I hope you don't mind asking you a question.
I've installed both icu.net and Icu4c.Win.Full.Lib

```
  <package id="icu.net" version="2.1.0-beta0017" targetFramework="net45" />
  <package id="Icu4c.Win.Full.Lib" version="56.1.5" targetFramework="net45" />
```

All dll's are being copied to the output directory:
icu,net.dll
icudt56.dll
icuin56.dll
icuio56.dll
icule56.dll
iculx56.dll
icutest55.dll
icutu55.dll
icuuc56.dll

The nugget packages created the proper .csproj configurations:
`  <Import Project="..\packages\Icu4c.Win.Full.Lib.56.1.5\build\Icu4c.Win.Full.Lib.targets" Condition="Exists('..\packages\Icu4c.Win.Full.Lib.56.1.5\build\Icu4c.Win.Full.Lib.targets')" />`

Nevertheless, I keep getting the following exception when using a RuleBasedBreakIterator:

```
System.IO.FileLoadException was unhandled by user code
  FileName=icuuc
  HResult=-2146232799
  Message=Can't load ICU library
  Source=icu.net
  StackTrace:
       at Icu.NativeMethods.LoadIcuLibrary(String libraryName) in c:\JenkinsSlaveHome\slave2\workspace\IcuDotNet-Win-any-master-release\source\icu.net\NativeMethods.cs:line 138
       at Icu.NativeMethods.get_IcuCommonLibHandle() in c:\JenkinsSlaveHome\slave2\workspace\IcuDotNet-Win-any-master-release\source\icu.net\NativeMethods.cs:line 90
       at Icu.NativeMethods.uloc_canonicalize(String localeID, IntPtr name, Int32 nameCapacity, ErrorCode& err) in c:\JenkinsSlaveHome\slave2\workspace\IcuDotNet-Win-any-master-release\source\icu.net\NativeMethods.cs:line 1423
       at Icu.Locale.GetString(GetStringMethod method, String localeId) in c:\JenkinsSlaveHome\slave2\workspace\IcuDotNet-Win-any-master-release\source\icu.net\Locale.cs:line 197
       at Icu.Locale..ctor(String localeId) in c:\JenkinsSlaveHome\slave2\workspace\IcuDotNet-Win-any-master-release\source\icu.net\Locale.cs:line 26
       at ReviewPortal.TextTools.TextMeasurer.WrapText(Line line, Font font) in C:\Projects\display-text-system\cjk-linebreaking\ReviewPortal\TextTools\TextMeasurer.cs:line 159
       at ReviewPortal.Web.Controllers.DisplayController.WrapText(IEnumerable`1 textContainers) in C:\Projects\display-text-system\cjk-linebreaking\ReviewPortal.Web\Controllers\DisplayController.cs:line 207
       at lambda_method(Closure , ControllerBase , Object[] )
       at System.Web.Mvc.ReflectedActionDescriptor.Execute(ControllerContext controllerContext, IDictionary`2 parameters)
       at System.Web.Mvc.ControllerActionInvoker.InvokeActionMethod(ControllerContext controllerContext, ActionDescriptor actionDescriptor, IDictionary`2 parameters)
       at System.Web.Mvc.Async.AsyncControllerActionInvoker.<BeginInvokeSynchronousActionMethod>b__39(IAsyncResult asyncResult, ActionInvocation innerInvokeState)
       at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncResult`2.CallEndDelegate(IAsyncResult asyncResult)
       at System.Web.Mvc.Async.AsyncControllerActionInvoker.EndInvokeActionMethod(IAsyncResult asyncResult)
       at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<InvokeActionMethodFilterAsynchronouslyRecursive>b__3d()
       at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<>c__DisplayClass46.<InvokeActionMethodFilterAsynchronouslyRecursive>b__3f()
  InnerException: 

```

Do you have any idea how to properly load the library ?
Cheers!

**Comment by funex** [comment]
Just an update:

I've used icu-dotnet source code instead of the nugget package to debug the library loading.
Line 151 - NativeMethods.cs

```
// libpath = "C:\\Projects\\display-text-system\\cjk-linebreaking\\ReviewPortal.Web\\bin\\icuuc56.dll"

var libPath = string.IsNullOrEmpty(_IcuPath) ? libName : Path.Combine(_IcuPath, libName);
handle = LoadLibrary(libPath);
```

The libPath is correct and the file exists, however for some reason LoadLibrary() just returns a Zero pointer and proceeds to try and load lower versions, which of course won't find.

Any ideas?

**Comment by funex** [comment]
Are you planning to have a nuget package to target both x86 and x64 ?

**Comment by ermshiperete** [comment]
> Indeed, if I compile to 32-bit it works.

How did you do that? I just tried and it didn't work for me.

> Are you planning to have a nuget package to target both x86 and x64 ?

Yes, if necessary we'll create one. So far there was no need for it since our apps are still 32-bit and no-one complained :smiley:

**Comment by funex** [comment]
I used a sample application to test a few things and toggled "Prefer 32-bit" under Platform target.

Btw, just for future reference, I was testing icu.net with a newer ICU4C version (5.8) and was having some problems loading the DLL's. 
I kept getting 126 error codes when trying to load icuuc58.dll, after some digging (dependency walker and process monitor) I noticed it was due to failing to locate its dependencies (icudt58.dll). Apparently, by default it will look for all directories assigned to the PATH environment variable.
So before loading the libraries I added a search order pointing to the dll's directory path:

```
[DllImport("kernel32.dll", SetLastError = true)]
private static extern bool SetDllDirectory(string dllDirectory);
```
